File: examples/cpp/player/agg/handlers/withdraw_handler.cpp
#include "withdraw_handler.hpp"
#include "angzarr/errors.hpp"
#include <chrono>

namespace player {
namespace handlers {

examples::FundsWithdrawn handle_withdraw(const examples::WithdrawFunds& cmd, const PlayerState& state) {
    // Guard
    if (!state.exists()) {
        throw angzarr::CommandRejectedError::precondition_failed("Player does not exist");
    }

    // Validate
    int64_t amount = cmd.has_amount() ? cmd.amount().amount() : 0;
    if (amount <= 0) {
        throw angzarr::CommandRejectedError::invalid_argument("amount must be positive");
    }
    if (amount > state.available_balance()) {
        throw angzarr::CommandRejectedError::precondition_failed("Insufficient funds");
    }

    // Compute
    int64_t new_balance = state.bankroll - amount;

    examples::FundsWithdrawn event;
    event.mutable_amount()->CopyFrom(cmd.amount());
    event.mutable_new_balance()->set_amount(new_balance);
    event.mutable_new_balance()->set_currency_code("CHIPS");

    auto now = std::chrono::system_clock::now();
    auto seconds = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
    event.mutable_withdrawn_at()->set_seconds(seconds);

    return event;
}

} // namespace handlers
} // namespace player

File: examples/csharp/Player/Agg/Handlers/WithdrawHandler.cs
using Google.Protobuf.WellKnownTypes;
using Angzarr.Client;
using Angzarr.Examples;

namespace Player.Agg.Handlers;

/// <summary>
/// Handler for WithdrawFunds command.
/// </summary>
public static class WithdrawHandler
{
    public static FundsWithdrawn Handle(WithdrawFunds cmd, PlayerState state)
    {
        // Guard
        if (!state.Exists)
            throw CommandRejectedError.PreconditionFailed("Player does not exist");

        // Validate
        var amount = cmd.Amount?.Amount ?? 0;
        if (amount <= 0)
            throw CommandRejectedError.InvalidArgument("amount must be positive");
        if (amount > state.AvailableBalance)
            throw CommandRejectedError.PreconditionFailed("Insufficient funds");

        // Compute
        var newBalance = state.Bankroll - amount;
        return new FundsWithdrawn
        {
            Amount = cmd.Amount,
            NewBalance = new Currency { Amount = newBalance, CurrencyCode = "CHIPS" },
            WithdrawnAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };
    }
}
